src/dotnet-core-uninstall/Shared/VSVersioning/VisualStudioSafeVersionsExtractor.cs

```diff
@@ -61,48 +61,52 @@ private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) A
                 .ToDictionary(key => key.Item1, value => value.Item2); 
 
-            var sdks = bundleList.Where(bundle => bundle.Version is SdkVersion);
-            if (sdks != null && sdks.Count() > 0)
+            if (macOSPreserveVSSdks)
             {
-                dividedBundles.Add(sdks, LocalizableStrings.MacSDKRequirementExplanationString);
+                var sdks = bundleList.Where(bundle => bundle.Version is SdkVersion) ?? [];
+                if (sdks.Any())
+                {
+                    dividedBundles.Add(sdks, LocalizableStrings.MacSDKRequirementExplanationString);
+                }
             }
 
             var remainingBundles = bundleList
-                .Where(bundle => !(bundle.Version is RuntimeVersion || bundle.Version is SdkVersion))
+                .Where(bundle => bundle.Version is not RuntimeVersion && !(bundle.Version is SdkVersion && macOSPreserveVSSdks))
                 .Concat(bundlesAboveLimit);
             return (dividedBundles, remainingBundles);
         }
 
-        private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) ApplyVersionDivisions(IEnumerable<Bundle> bundles)
+        private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) ApplyVersionDivisions(IEnumerable<Bundle> bundles, bool macOSPreserveVSSdks = false)
         {
             if (RuntimeInfo.RunningOnWindows)
             {
                 return ApplyWindowsVersionDivisions(bundles);
             }
             else
             {
-                return ApplyMacVersionDivisions(bundles);
+                return ApplyMacVersionDivisions(bundles, macOSPreserveVSSdks);
             }
         }
 
-        public static IEnumerable<Bundle> GetUninstallableBundles(IEnumerable<Bundle> bundles)
+        public static IEnumerable<Bundle> GetUninstallableBundles(IEnumerable<Bundle> bundles, bool macOSPreserveVSSdks = false)
         {
-            var required = new List<Bundle>();
-            var (bundlesByDivisions, remainingBundles) = ApplyVersionDivisions(bundles);
+            var protectedBundles = new List<Bundle>();
+            var (bundlesByDivisions, remainingBundles) = ApplyVersionDivisions(bundles, macOSPreserveVSSdks);
 
             foreach (IEnumerable<Bundle> band in bundlesByDivisions.Keys)
             {
-                required.Add(band.Max());
+                protectedBundles.Add(band.Max());
             }
 
-            required = required.Concat(remainingBundles.Where(bundle => bundle.Version.SemVer >= UpperLimit)).ToList();
+            protectedBundles = protectedBundles.Concat(remainingBundles.Where(bundle => bundle.Version.SemVer >= UpperLimit)).ToList();
 
-            return bundles.Where(b => !required.Contains(b));
+            // Return all bundles that are not protectedBundles
+            return bundles.Except(protectedBundles);
```

Comment:
I think this is correct, but I got burned at one point making a change like this because this looks equivalent but isn't...the order doesn't matter, and you don't need duplicates, right?

Reply:
Yes. But to be safe I can revert